test_plans/group_functions.asciidoc

```diff
@@ -0,0 +1,276 @@
+* `int`
+* `float`
+* `long long`
+
+For FULL conformance add:
+
+* `signed char`
+* `unsigned char`
+* `short`
+* `unsigned short`
+* `unsigned int`
+* `long`
+* `unsigned long`
+* `unsigned long long`
+* `bool`
+
+In addition, if the device has aspect::fp64:
+
+* `double`
+
+In addition, if the device has aspect::fp16:
+
+* `half`
+
+If function accepts another types, the following list of types is added:
+
+* `vec<2, int>`
+* `vec<16, unsigned long long>`
+* `marray<5, float>`
+* `marray<7, short>`
+* `struct { int a; bool b; }`
+
+Pointers are taken from the same list of types.
+
+`bool Predicate(T)` functions are functions checking the ranges of `T` itself or
+the first element of `vec, marray, struct` type `T`. They are choosen in such a way that
+functions using them should return both `true` and `false` in different tests.
+
+`binary_op` used in scans and reductions are addition and maximum of `T` itself or
+the first element of `vec, marray, struct` type `T`.
+
+=== Data for the functions
+
+Data for the functions are sequences of 1, 2, 3, etc. corresponding to increasing `linear_id`.
+
+=== Execution
+
+Functions will be executed by a group using `sycl::parallel_for` with `nd_range<Dims>`
+corresponding to one group of maximal group size available on device with `Dims = 1,2,3`
+(by querying `info::device::max_work_item_sizes<Dims>` and `info::device::max_work_group_size`).
+Sub-group variants will execute functions on the first subgroup of this large group only.
+
+=== Traits
+
+==== is_group
+
+For both `group` and `sub_group` template parameter T check that 
+`is_group<T>` is `std::true_type`. For some other template parameter T
+check that `is_group<T>` is `std::false_type`.
+
+==== is_group_v
+
+For both `group` and `sub_group` template parameter T check that 
+`is_group_v<T>` is `true`. For some other template parameter T
+check that `is_group_v<T>` is `false`.
+
+=== Functions
+
+==== group_broadcast
+
+For each of the following use-cases:
+
+* using `T group_broadcast(Group g, T x);`
+* using `T group_broadcast(Group g, T x, Group::linear_id_type local_linear_id);`
+* using `T group_broadcast(Group g, T x, Group::id_type local_id);`
+
+for some scalar, vector, marray, and user defined trivially copyable
+types `T` and both `group` and `sub_group` check returned type and value.
+
+==== group_barrier
+
+For each possible `memory_scope` wider than `Group::fence_scope` in the list
+
+* `sycl::memory_scope::sub_group`
+* `sycl::memory_scope::work_group`
+* `sycl::memory_scope::device`
+* `sycl::memory_scope::system`
+
+check that invocation of `void group_barrier(Group g, memory_scope fence_scope)`
+for both `group` and `sub_group` behaves as expected: when each of workitems writes 1
+into zero-initialized local memory array (for `sub_group` and `work_group` memory scope)
+or zero-initialized global memory array (for `device` and `system` memory scope)
+after barrier all array values red by workitems in reverse order are equal to 1.
```

Comment:
I think you will need to write the test so that it checks that all these memory scopes are supported by using `info::device::atomic_fence_scope_capabilities`.  The specification says that only `memory_scope::work_group` is required to be supported.  The test should skip any memory scopes that are not supported according to that query.

Adding @pennycook here ... should the spec also guarantee that `memory_scope::sub_group` is supported?  The default memory  scope for `group_barrier<sub_group>()` is `memory_scope::sub_group`.  It seems weird to have a default value that might not be supported by the device.  In any case, isn't it trivial to support `memory_scope::sub_group` if you are required to support `memory_scope::work_group`?

Reply:
Yes, you're right, it should be tested before.

Reply:
@gmlueck, @Pennycook, should I wait for @bashbaug to reply before merging this PR?

Reply:
Note: I think this is one of the places where SYCL and OpenCL can be different, if desired, though we'll need some care if they are different.

In OpenCL in general, the "subgroup" memory scope is only required when subgroups are supported.

* In OpenCL C it only exists when subgroups are supported ([link](https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_C.html#memory-scope) - "Requires support for OpenCL C 3.0 or newer and the __opencl_c_subgroups feature.").
* In SPIR-V, the subgroup memory scope is only valid when subgroups are supported ([link](https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_Env.html#validation-rules) - "For OpenCL 2.1, OpenCL 2.2, or OpenCL 3.0 devices supporting Subgroups (where CL_DEVICE_MAX_NUM_SUB_GROUPS is not 0)."  See note below...).
* This means that a device that does not support subgroups does not need to handle a subgroup scope memory fence, say by "upgrading" it to the wider work-group scope.

If SYCL chooses to require the subgroup scope when the work-group scope is also supported it's fine to do so, but it will be SYCL's responsibility to generate valid SPIR-V in this case, specifically by generating a work-group fence instead of a subgroup fence if/when subgroup fences aren't supported.

I agree with most of @pennycook's suggested clarifications above, especially that we should require that the subgroup scope is supported when subgroups are supported.  This is debatably already true, but it is not stated explicitly.

When we do this I also think it makes sense to describe the SPIR-V requirements in terms of the atomic fence capabilities query, rather than subgroups generically.  Again, this is debatably already true, but I think it's a bit clearer and consistent to describe the requirements in this way.

Reply:
I think we can merge this test plan before this conversation is resolved.  The only impact on the test plan is whether the test can assume that `memory_scope::sub_group` is supported or if it must check to see if the device supports this memory scope.

I suggest that the CTS follow the existing wording of the SYCL spec, which does *not* guarantee that `memory_scope::sub_group` is supported.

I have opened the following issue to continue the conversation about whether the SYCL spec should be changed:

https://gitlab.khronos.org/sycl/Specification/-/issues/639